codingStandards/AdditionalGuidelines-Java.md

```diff
@@ -0,0 +1,98 @@
+# Additional Guidelines - Java
+
+## Argument validation
+
+Validating arguments in _every_ method can be an unnecessary cost.
+Following guidelines try to reduce the argument validation cost without losing its benefits. 
+
+1. **Validate an argument only if an invalid value can result in an _undesirable_ yet _non-failure_ behavior _directly_ 
+   within the method.**
+
+   In the example below, `name` should be validated because an invalid `name` can result in invalid data in 
+   the system from that point onwards without indicating that something went wrong.
+   
+   ```java
+   void setName(String name){
+       this.name = name;
+   }
+   ```
+   
+   However, `name` in the example below need not be validated because it is not being used _directly_ in the method.
+   
+   ```java
+   void setName(String name){
+       this.saveName(name);
+   }
+   ```
+   
+   In the example below, `key` need not be validated against `null` because a `null` value is going to create a 
+   detectable failure (i.e. a `NullPointerException`) anyway.
+   
+   ```java
+   boolean isPersonKey(String key){
+       return key.contains("P");
+   }
+   ```
+   
+   Furthermore, place the argument validation code immediately above the first line of code it tries to defend, so that
+   the reader can easily deduce the reason for argument validation.
```

Comment:
Indeed @acjh , validating arguments upfront seems like the natural thing to do and I used to do it that way too. However, now that the decision to validate depends on a specific circumstances in the code, it seems safer to put the validation near the 'reason for validation' so that if the 'reason for validation' changes, the validation code too can be changed accordingly. For example, if the 'reason for validation' code is extracted to another method, the validation code can go with it. But I don't feel strongly about it; it's just something I cooked up (and yes, I too could not find a good reference to derive from).